velox/common/memory/MemoryArbitration.h

```diff
@@ -0,0 +1,187 @@
+/// memory limit. Sapphire needs a shared memory arbitration policy to achieve
+/// dynamic memory sharing across queries so that one query can use more than
+/// its fair share during the time that the system has free available memory and
+/// the excess memory can be reclaimed back on-demand later through memory
+/// reclaim mechanism such as spilling techniques.
+class MemoryArbitrator {
+ public:
+  /// Defines the kind of memory arbitrators.
+  enum class Kind {
+    /// Used by Prestissimo to enforce fixed size memory isolation across
+    /// queries running on the same node.
+    kFixed,
+    /// Used by Sapphire to achieve dynamic memory sharding across quries
+    /// running on the same node.
+    kShared,
+  };
+
+  struct Config {
+    Kind kind;
+    /// Specifies the total memory capacity in bytes.
+    int64_t capacity;
+  };
+  static std::unique_ptr<MemoryArbitrator> create(const Config& config);
+
+  Kind kind() const {
+    return kind_;
+  }
+  static std::string kindString(Kind kind);
+
+  virtual ~MemoryArbitrator() = default;
+
+  /// Invoked by the memory manager to reserve the initial memory on a memory
+  /// pool creation.
+  virtual int64_t reserveMemory(int64_t targetBytes) = 0;
+
+  /// Invoked by the memory manager to release all the reserved memory from a
+  /// memory pool on its destruction.
+  virtual void releaseMemory(MemoryPool* releasor) = 0;
+
+  /// Indicates if this arbitrator allows to grow a memory pool's limit
+  /// dynamically.
+  virtual bool canGrow() = 0;
+
+  /// Invoked by the memory manager to grow a memory pool's limit. 'requestor'
+  /// is the memory pool to request to grow. 'candidates' is a list of root
+  /// pools to participate in the memory arbitration. The memory arbitrator
+  /// picks up a number of pools to either shrink its memory limit or reclaim
+  /// its used memory to free up enough memory for 'requestor' to grow.
+  /// Different arbitrators use different policies to select the candidate
+  /// pools. For instance, the arbitrator used by Sapphire select candidates
+  /// which have used more than its fair shares. 'targetBytes' is the number of
+  /// bytes to grow for 'requestor'.
+  ///
+  /// NOTE: the memory manager keeps 'candidates' valid during the arbitration
+  /// processing.
+  virtual bool growMemory(
+      MemoryPool* requestor,
+      const std::vector<MemoryPool*>& candidates,
+      uint64_t targetBytes) = 0;
+
+  /// The internal execution stats of the memory arbitrator.
+  struct Stats {
+    /// The number of arbitration requests.
+    uint64_t numRequests{0};
+    /// The number of arbitration request failures.
+    uint64_t numFailures{0};
+    /// The number of arbitration requests that has been queued waiting for
+    /// execution.
+    uint64_t numQueuedRequests{0};
+    /// The sum of all the arbitration request queue times in microseconds.
+    uint64_t queueTimeUs{0};
+    /// The number of arbitration runs.
```

Comment:
What's the difference between "run" and "request"?

Reply:
I just request and it is left out from an early version of POC.